lib/ruby_lsp/requests/base_request.rb

```diff
@@ -1,24 +1,34 @@
-# typed: true
+# typed: strict
 # frozen_string_literal: true
 
 module RubyLsp
   module Requests
     # :nodoc:
     class BaseRequest < SyntaxTree::Visitor
+      extend T::Sig
+      extend T::Helpers
+      extend T::Generic
+
+      Response = type_template { { upper: T.untyped } }
+
+      abstract!
+
+      sig { overridable.params(document: Document).returns(Response) }
       def self.run(document)
         new(document).run
       end
 
+      sig { params(document: Document).void }
       def initialize(document)
         @document = document
 
         super()
       end
 
-      def run
-        raise NotImplementedError, "#{self.class}#run must be implemented"
-      end
+      sig { abstract.returns(T.untyped) }
```

Comment:
Isn't it always returning a `Response`?

```suggestion
      sig { abstract.returns(Response) }
```

Reply:
Yes, but Sorbet doesn't allow us to use a `type_template` in instance methods. Because the `run` method must be overridden by child classes, we get the correct typing without the need of adding a second generic as a `type_member`.

Reply:
I am not sure if `self.run` methods are adding any value to our code, to be honest. We instantiate and run these request classes in one location in our codebase and a few locations at most in our tests. Is it worth going through a `self.run` method that is hard to type?

Reply:
I agree, `self.run` is not providing any value. I removed those and removed the generics in the latest commit.

The reason `CodeActions` doesn't inherit from `BaseRequest` is because all it does is run `Diagnostics` and then filter the results.

Reply:
Could we not have the generic still but as a `type_member` now so that we can use it as the return type of `run`?

Reply:
> The reason CodeActions doesn't inherit from BaseRequest is because all it does is run Diagnostics and then filter the results.

Still, doesn't it implement the same interface than `BaseRequest`? What makes it incompatible with `BaseRequest`?